### PR TITLE
Improved handling of power calculations when effect-size = 0

### PR DESCRIPTION
**Pull request by mikekaminsky**
Warn the user that no amount of power can detect an effect size of 0

First time contributor, open to suggestions! Also I had a lot of trouble with local installation and wasn't able to get tests to run locally. Seeing what happens with Travis....

### CODE REVIEW

**Comment by josef-pkt** [comment]
It's good to handle the corner case.

The question is whether this should return nan or raise instead.
Strictly speaking inf is not the answer if the effect size == 0, although it is the limiting case as effect size approaches zero.
This is different as limit or local alternative from cases like #4022, where there is no possible case in the neighborhood that would have a solution, not even with inf.

Another issue: which warning, not clear yet, this is only the second case

`proportion._power_ztost` uses HypothesisTestWarning, where the zero power case is not obvious from the user input

```
    if np.any(k_low > k_upp):   #vectorize
        import warnings
        warnings.warn("no overlap, power is zero", HypothesisTestWarning)
```

**Comment by mikekaminsky** [comment]
@josef-pkt thanks for the comments. Upon reflection, I like raising here as I suspect that's the best way to force the user to actually *think* about what they're trying to do.

**Comment by josef-pkt** [comment]
just another case that occured to me (not checked yet)

If we are looking for either alpha or beta/power, then there is a solution, i.e. the rejection rate is the same under the null and under the alternative. IIRC, I used this to check that power is equal to alpha if the alternative is the same as the null hypothesis.

**Comment by mikekaminsky** [comment]
@josef-pkt I'm not sure I follow that. 

TBH my understanding of power analysis is more intuitive than formal so I'm happy to follow your guidance on implementation, but it doesn't make sense to me that power can be defined for an effect-size of zero or that there would be any alpha you could choose to detect such an effect.

**Comment by josef-pkt** [comment]
Think of power in terms of the probability of rejecting the Null for a given effect size.
Under the Null the rejection rate is given by alpha.
Under a alternative the rejection rate is the power given the effect size.

If we plot rejection probability as a function of effect size, then it starts at alpha e.g. equal to 0.05 when the effect size is zero and continuously increases as the (absolute value of the) effect size increases.

So "power" is well defined at zero effect size, but it is the same as alpha,

File: statsmodels/stats/tests/test_power.py [hunk 1]
```diff
@@ -734,6 +734,9 @@ def test_power_solver():
     assert_raises(ValueError, nip.solve_power, None, nobs1=1600, alpha=0.01,
                   power=0.005, ratio=1, alternative='larger')
 
+    assert_raises(ValueError, nip.solve_power, None, effect_size=0, alpha=0.01,
```
Comment (josef-pkt) [inline statsmodels/stats/tests/test_power.py]: I guess this needs to use `nobs1=None` instead of just the None

**Comment by mikekaminsky** [comment]
Fascinating. [Aligns with Wikipedia](https://en.wikipedia.org/wiki/Statistical_power#Interpretation).

@josef-pkt I think my implementation is now correct -- I return `alpha` when looking for `power` (and vice versa) when `effect_size = 0`.

File: statsmodels/stats/power.py [hunk 2]
```diff
@@ -219,6 +219,15 @@ def solve_power(self, **kwds):
             del kwds['power']
             return self.power(**kwds)
 
+        if kwds['effect_size'] == 0:
+            if key == 'nobs1':
+                raise ValueError('Cannot detect an effect-size of 0. Try changing your effect-size.')
+            print('Warning: Effect size of 0 detected')
```
Comment (josef-pkt) [inline statsmodels/stats/power.py]: We don't use print in any code (maybe with the exception of the optimizaters)
Either delete or convert to warnings.warn

**Comment by josef-pkt** [comment]
looks good to me except for the one print.

Can you rebase and squash? Otherwise I will do it.

File: statsmodels/stats/power.py [hunk 3]
```diff
@@ -219,6 +219,15 @@ def solve_power(self, **kwds):
             del kwds['power']
             return self.power(**kwds)
 
+        if kwds['effect_size'] == 0:
+            if key == 'nobs1':
```
Comment (josef-pkt) [inline statsmodels/stats/power.py]: It's better to put the raise as default in a final `else`

File: statsmodels/stats/power.py [hunk 4]
```diff
@@ -219,6 +219,15 @@ def solve_power(self, **kwds):
...
+                raise ValueError('Cannot detect an effect-size of 0. Try changing your effect-size.')
+            print('Warning: Effect size of 0 detected')
+            if key == 'power':
+                return alpha
+            if key == 'alpha':
```
Comment (josef-pkt) [inline statsmodels/stats/power.py]: elif key == 'alpha'

and the the else with raise

Comment (josef-pkt) [reply statsmodels/stats/power.py]: Just saw: `elif` is not needed because it `returns` after if

**Comment by josef-pkt** [comment]
One more reshuffling
If the raise is at the end, then it will raise for all cases that are not solving for alpha and power.
E.g. we also want to raise if we search for `ratio`

File: statsmodels/stats/power.py (hunk 2 above)
Comment (mikekaminsky) [reply statsmodels/stats/power.py]: I just copied [from this line](https://github.com/statsmodels/statsmodels/blob/master/statsmodels/stats/power.py#L240)
```
print('Warning: using default start_value for {0}'.format(key))
```

I'll convert both to real warnings

Comment (josef-pkt) [reply statsmodels/stats/power.py]: converting also the one from line 240 is good.
(I think that one should never show up in user applications. Based on the comment and AFAIR, this was a reminder for developers that they (mainly I) should add the method to new subclasses instead of relying on default behavior.)

**Comment by mikekaminsky** [comment]
@josef-pkt I did the reshuffling and squashed it down. I left the change of the unrelated print statement in a separate commit in case we want to revert that

File: statsmodels/stats/power.py (hunk 2 above)
Comment (mikekaminsky) [reply statsmodels/stats/power.py]: Got it -- I made the change for the sake of future developers :)

**Comment by josef-pkt** [comment]
@mikekaminsky  Thanks, looks very good. The warning is better than print even in developer-only code paths.

I will merge it when the test CI report back green.
